crates/next-core/src/app_structure.rs

```diff
@@ -0,0 +1,316 @@
+use std::collections::HashMap;
+
+use anyhow::{bail, Result};
+use turbo_tasks::{
+    primitives::{StringVc, StringsVc},
+    CompletionVc, ValueToString,
+};
+use turbo_tasks_fs::{
+    DirectoryContent, DirectoryEntry, File, FileContentVc, FileSystemEntryType, FileSystemPathVc,
+};
+use turbopack_core::issue::{Issue, IssueSeverity, IssueSeverityVc, IssueVc};
+use turbopack_dev_server::source::specificity::SpecificityVc;
+
+use crate::{
+    app_render::{LayoutSegment, LayoutSegmentVc, LayoutSegmentsVc},
+    embed_js::wrap_with_next_js_fs,
+    next_config::NextConfigVc,
+};
+
+#[turbo_tasks::value]
+pub enum AppStructureItem {
+    Page {
+        segment: LayoutSegmentVc,
+        segments: LayoutSegmentsVc,
+        url: FileSystemPathVc,
+        specificity: SpecificityVc,
+        page: FileSystemPathVc,
+    },
+}
+
+#[turbo_tasks::value_impl]
+impl AppStructureItemVc {
+    #[turbo_tasks::function]
+    pub async fn routes_changed(self) -> Result<CompletionVc> {
+        match *self.await? {
+            AppStructureItem::Page { url, .. } => url.await?,
+        };
+        Ok(CompletionVc::new())
+    }
+}
+
+#[turbo_tasks::value]
+pub struct AppStructure {
+    pub directory: FileSystemPathVc,
+    pub item: Option<AppStructureItemVc>,
+    pub children: Vec<AppStructureVc>,
+}
+
+#[turbo_tasks::value_impl]
+impl AppStructureVc {
+    #[turbo_tasks::function]
+    pub async fn directory(self) -> Result<FileSystemPathVc> {
+        Ok(self.await?.directory)
+    }
+
+    #[turbo_tasks::function]
+    pub async fn routes_changed(self) -> Result<CompletionVc> {
```

Comment:
Explain somewhere that this `routes_changed` method is a way to recursively watch for route changes, by marking leaf route URLs as dependencies.